dap: change how noDebug launch request is served

hyangah opened the pull request:
PR #2400 added support for noDebug launch requests - that was done
by directly starting the target program using os/exec.Cmd and blocking
the launch request indefinitely until the program terminates or
is stopped with a disconnect request (when dlv dap can support it).
This PR updates the logic slightly - instead of skipping the launch response,
we now return the launch response. But still we don't send 'initialized' event
hoping that the client not to send any debug-related requests.

Still the handler goroutine will block in the onLaunchRequest handler until
the debugged program terminates.

Updates #2318

hyangah commented:
cc @polinasok

hyangah commented:
One drawback of this approach is that this will not allow users to run a program built without debug info. (VS Code currently allows to combine `exec` mode launch tasks with arbitrary preLaunch/postLaunch tasks). 
If there is a user of such arrangement, we can bring back `os/exec.Command`-based execution here in the future. For now, I hope this is simpler and minimizes further interruption to other ongoing work (i.e. async requests).

aarzilli commented:
Note that, in addition to not working when programs aren't built with debug info, there are other notable differences:

* on macOS and windows asyncpreempt will be disabled
* on darwin/arm64 ASLR will be disabled
* on macOS the use will sometimes be prompted for their password
* on macOS all versions of xcode that do not have the patch I submitted last year will not be able to run programs that `recover()` from a nil pointer dereference (and in general nil pointer dereferences will behave differently) -- I do not know if any released version of xcode contains the workaround
* running on architectures not supported by delve (for example windows/386) will not be possible

hyangah commented:
Thanks @aarzilli for the information. Some of them sound like a huge drawback.
Then, an alternative I can think of is to reenable the os/exec.Cmd logic back, but call its `Run` when ConfigurationDone message is received. I hope this doesn't interfere too much with the other planned work.

polinasok reviewed (commented):
Please add a reference to #2318 to the description. 

As for the new design, I found your original approach quite clever in that it disabled all but restart/disconnect controls in the UI. I suspect that this new one enables Pause even though we don't support it. The idea that I had to refine the first version was to send the launch response back, but not to send the initialize event, which is what signals to the editor that the debug adapter is ready to accept breakpoints and such. So if you skip it, then you should not get any such requests, including ConfigurationDone and ThreadsRequest.

We should still make sure that all theses debug request handlers fail gracefully in the absense of debugging. The evaluate request should already be responding with debugger-nil error. But we might want to make sure all others do the same and explicitly check for nil debugger. Even if the UI is disabled, one could trigger something unexpected with keyboard shortcuts.

polinasok commented on service/dap/server_test.go:
```diff
@@ -2377,44 +2377,37 @@ func TestLaunchRequestDefaults(t *testing.T) {
 
 func TestLaunchRequestDefaultsNoDebug(t *testing.T) {
```
Oh, and please remove "Defaults" from the name. Just like the duplicate tests below that you removed, it's an artifact from copying another test that tested for various defaults of the args in the launch arguments.

hyangah replied:
Done

polinasok commented on service/dap/server_test.go:
```diff
@@ -2377,44 +2377,37 @@ func TestLaunchRequestDefaults(t *testing.T) {
 
 func TestLaunchRequestDefaultsNoDebug(t *testing.T) {
 	runTest(t, "increment", func(client *daptest.Client, fixture protest.Fixture) {
-		runNoDebugDebugSession(t, client, "launch", func() {
```
how about runNoDebugSession? Two debugs are odd.

hyangah replied:
Done

polinasok commented on service/dap/server_test.go:
```diff
@@ -2377,44 +2377,37 @@ func TestLaunchRequestDefaults(t *testing.T) {
-		runNoDebugDebugSession(t, client, "launch", func() {
+		runNoDebugDebugSession(t, client, func() {
 			client.LaunchRequestWithArgs(map[string]interface{}{
 				"noDebug": true,
```
Let's add some bad testing to TestBadLaunchRequests.

hyangah replied:
WDYT

hyangah commented on service/dap/server.go:
```diff
@@ -549,75 +550,58 @@ func (s *Server) onLaunchRequest(request *dap.LaunchRequest) {
-	// will end the configuration sequence with 'configurationDone'.
 	s.send(&dap.InitializedEvent{Event: *newEvent("initialized")})
 	s.send(&dap.LaunchResponse{Response: *newResponse(request.Request)})
+	// TODO: if the client doesn't have supportConfigurationDone capability,
```
you're right. it's my misunderstanding. Removed.

hyangah commented on service/dap/server.go:
```diff
@@ -71,10 +70,9 @@ type Server struct {
 	// noDebugProcess is set for the noDebug launch process.
-	noDebugProcess *exec.Cmd
+	noDebugProcess           *exec.Cmd
+	noDebugProcessTerminated bool
```
Placed back the mutex and removed bool.
I don't remember what made me to add the termination flag (there was some strange case), but thinking again, the underlying issue is the two goroutines racing, not the termination state.

hyangah commented on service/dap/server.go:
```diff
@@ -549,75 +550,58 @@ func (s *Server) onLaunchRequest(request *dap.LaunchRequest) {
-			return
+	noDebug := false
+	if noDebugArg, ok := request.Arguments["noDebug"]; ok {
+		if v, ok := noDebugArg.(bool); ok && v { // noDebug == true
```
done.

hyangah commented on service/dap/server.go:
```diff
@@ -549,75 +550,58 @@ func (s *Server) onLaunchRequest(request *dap.LaunchRequest) {
-			FailedToLaunch, "Failed to launch", err.Error())
-		return
+	if noDebug {
+		if s.noDebugProcess, err = s.launchWithoutDebug(program, targetArgs, s.config.Debugger.WorkingDir); err != nil {
```
now renamed to startNoDebugProcess, and there is stopNoDebugProcess.

hyangah commented on service/dap/server.go:
```diff
@@ -652,14 +636,16 @@ func (s *Server) onDisconnectRequest(request *dap.DisconnectRequest) {
 			s.log.Error(err)
 		}
-	} else {
-		s.stopNoDebugProcess()
```
I moved to Stop previously to match its debugger.Detach call. But now I am not sure about the teardown logic in general.
Changed to call it from both Stop and onDisconnectRequest for now (checking on s.noDebugProcess and setting it back to null will prevent double process kill)

hyangah commented on service/dap/server.go:
```diff
@@ -549,75 +550,58 @@ func (s *Server) onLaunchRequest(request *dap.LaunchRequest) {
+func (s *Server) launchWithoutDebug(program string, targetArgs []string, wd string) (*exec.Cmd, error) {
 	s.log.Println("Running without debug: ", program)
 
 	cmd := exec.Command(program, targetArgs...)
```
Updated

hyangah commented:
> Please add a reference to #2318 to the description.
> 
> As for the new design, I found your original approach quite clever in that it disabled all but restart/disconnect controls in the UI. I suspect that this new one enables Pause even though we would not support it even with an async implementation. The idea that I had to refine the first version was to send the launch response back, but not to send the initialize event, which is what signals to the editor that the debug adapter is ready to accept breakpoints and such. So if you skip it, then you should not get any such requests, including ConfigurationDone and ThreadsRequest.

Thanks for the suggestion. That was my initial attempt (before sending the PR) and I changed my mind because I hoped to match in what handler the request blocks in when the program is running (i.e. ConfigurationDone request handler currently). I hoped matching noDebug and debug cases as much as possible could help classifying what request handlers can be potentially blocking,  but it sounds like that's not a huge deal and blocking from the launch request handling is ok. Now I changed my mind again and implemented what you described.

hyangah commented:
The test failure (on TeamCity/mac) seems unrelated to this change
```
------- Stdout: -------
=== RUN   TestContinue
dlv_test.go:241: API server listening at: 127.0.0.1:40573
dlv_test.go:241: hello world!
dlv_test.go:250: error detaching from headless instance: unexpected EOF
--- FAIL: TestContinue (1.23s)
```

aarzilli commented:
> The test failure (on TeamCity/mac) seems unrelated to this change

It is.

polinasok commented on service/dap/server.go:
```diff
@@ -550,58 +550,66 @@ func (s *Server) onLaunchRequest(request *dap.LaunchRequest) {
-	var err error
-	if noDebug {
-		if s.noDebugProcess, err = s.launchWithoutDebug(program, targetArgs, s.config.Debugger.WorkingDir); err != nil {
+	if noDebug, _ := request.Arguments["noDebug"].(bool); noDebug {
```
Isn't it the recommended practice to check err/ok instead of relying on the default/value not getting populated when an operation fails? Or does that not apply to scalars?
noDebug, ok := request.Arguments["noDebug"].(bool); ok && noDebug {...}

aarzilli replied:
I'm not aware of anyone recommending this.

polinasok replied:
Good to know. I remember this coming up in some code review in my early Go days, but it was probably about structs, not scalars.

hyangah replied:
I am using this multi-value returning type assertion here to avoid panic.

According to the language spec, when the type assertion fails, the value is a zero value (false, 0, empty struct). In this case, zero value (false) makes sense, so I don't think this form is violating the spec or any style suggestion. (for example, https://golang.org/src/net/http/server.go#L3147) When accessing a struct field, checking ok may be useful when accessing the zero value's field isn't desirable. Even when accessing a scalar field, checking ok may be necessary when distinction between zero value vs uninitialized value is necessary. All depend on the context.

For errors, yes, it's strongly recommended not to discard error (but as always, there are exceptions).

polinasok commented on service/dap/server.go:
```diff
@@ -550,58 +550,66 @@ func (s *Server) onLaunchRequest(request *dap.LaunchRequest) {
+
+		// Then, block until the program terminates or is stopped.
+		if err := cmd.Wait(); err != nil {
+			s.log.Errorf("program exited: %v", err)
```
Based on your other CL, this should be Debug logging now :)

hyangah replied:
Done.

polinasok commented on service/dap/server.go:
```diff
@@ -550,58 +550,66 @@ func (s *Server) onLaunchRequest(request *dap.LaunchRequest) {
+	}
+
+	var err error
+	if s.debugger, err = debugger.New(&s.config.Debugger, s.config.ProcessArgs); err != nil {
```
does err := not work? (to avoid "var err error")

aarzilli replied:
`:=` only works with variable names on the left.

polinasok replied:
Ah, I see, "non-name ... on left side of :=". That's unfortunate.

hyangah replied:
The offical name of := is a short variable declaration, and 's.debugger' isn't a variable.

polinasok commented on service/dap/server.go:
```diff
@@ -550,58 +550,66 @@ func (s *Server) onLaunchRequest(request *dap.LaunchRequest) {
+
+		stopped := false
+		s.mu.Lock()
+		stopped = s.noDebugProcess == nil // if it was stopped, this should be nil.
```
If you set it to nil only here, you could check Cmd.ProcessState here, couldn't you? And that would simplify the logic in stopNoDebugProcess too a bit.

hyangah replied:
The program can terminate and we end up here because 1) the program exited by itself - due to normal termination or due to other signals (segfault, ...), or 2) the program was killed by Kill from stopNoDebugProcess. I don't see how to distinguish the two cases cleanly with Cmd.ProcessState when the program was killed by external processes.  In case of 2, we don't want to issue TerminatedEvent. Feel free to revisit/update it when the teardown logic & async request serving is settled down.

polinasok commented on service/dap/server.go:
```diff
@@ -637,6 +645,8 @@ func (s *Server) onDisconnectRequest(request *dap.DisconnectRequest) {
 			s.log.Error(err)
 		}
 	}
+	s.stopNoDebugProcess()
```
I know you safeguard against this being nil in the function, but I think the code was more clear when you had an else here. Either we have a debugger or a noDebug process.

hyangah replied:
done.
Strictly speaking, s.debugger != nil doesn't necessary mean noDebug mode (s.debugger could be in uninitialized state).

polinasok replied:
Since debugger is a pointer and pointers have nil default values in Go (unlike C, as far as I know), you must be referring to a case of exiting early before neither debugger nor process have been launched. The else could be else-if instead or in addition to doing the check in stopNoDebugProcess()

polinasok commented on service/dap/server_test.go:
```diff
@@ -2713,6 +2706,14 @@ func TestBadLaunchRequests(t *testing.T) {
 
 		client.LaunchRequestWithArgs(map[string]interface{}{"mode": "debug", "program": fixture.Source, "buildFlags": "bad flags"})
 		expectFailedToLaunchWithMessage(client.ExpectErrorResponse(t), "Failed to launch: Build error: exit status 1")
+		client.LaunchRequestWithArgs(map[string]interface{}{"mode": "debug", "program": fixture.Source, "noDebug": true, "buildFlags": "bad flags"})
```
you could also add a test above where noDebug is something other than a bool

hyangah replied:
noDebug other than a bool will be ignored (not a bad request) and the launch requests should work as if it wasn't set. I see the other way - failing on non-bool type value - isn't wrong either. I am not too worried about it right now. If you want, feel free to add extra check and make it to the bad request case.
Added to the launch default test for now.

polinasok replied:
Sure, looks good. I just meant we needed a test "documenting" behavior, whatever it tis.  
I did the same thing for a number of other flags - if they are bad, don't generate a launch error, but quietly use a default. Suzy and I had a discussion about this because she wanted to use the bad launch approach:
https://github.com/go-delve/delve/pull/2379#issuecomment-806053504

polinasok commented on service/dap/server_test.go:
```diff
@@ -2713,6 +2706,14 @@ func TestBadLaunchRequests(t *testing.T) {
+		client.LaunchRequestWithArgs(map[string]interface{}{"mode": "debug", "program": fixture.Source, "noDebug": false, "wd": "dir/invalid"})
+		expectFailedToLaunch(client.ExpectErrorResponse(t)) // invalid directory, the error message is system-dependent.
+		client.LaunchRequestWithArgs(map[string]interface{}{"mode": "debug", "program": fixture.Source, "noDebug": true, "wd": "dir/invalid"})
+		expectFailedToLaunch(client.ExpectErrorResponse(t)) // invalid directory, the error message is system-dependent.
```
good new tests cases, thanks!

polinasok commented on service/dap/server.go:
```diff
@@ -550,58 +550,66 @@ func (s *Server) onLaunchRequest(request *dap.LaunchRequest) {
-		if s.debugger, err = debugger.New(&s.config.Debugger, s.config.ProcessArgs); err != nil {
-			s.sendErrorResponse(request.Request, FailedToLaunch, "Failed to launch", err.Error())
-			return
+		// Skip 'initialized' event.
```
I think it would be good to document here that by doing so we avoid debug requests.

hyangah replied:
Done.

polinasok commented on service/dap/server.go:
```diff
@@ -70,9 +71,10 @@ type Server struct {
 	// args tracks special settings for handling debug session requests.
 	args launchAttachArgs
 
+	mu sync.Mutex
```
Let's call this noDebugProcessMux since we might end up with more locks and will need to differentiate.

polinasok replied:
Based no the comment in the other PR, I understand now what your intension was here. Please disregard this comment.

derekparker reviewed (approved):
LGTM
